labd/mach-composer
# Small maintenance

Some small improvements in the build pipeline: added internet-recommended build flags and couple of new goreleaser steps to improve packaging. New linting checks + improvements in the code organization (in accordance with https://github.com/golang-standards/project-layout)

## Changes

### Test pipeline

- Renamed to linux to reflect the tested platform name in the actions UI.
- Do not output artifacts for all platforms in the test pipeline, only run `make run`.
- Make test pipeline run on pull_requests as well. This should be cheap without the snapshots build.
- Use 1.18.x as golang version.
- Use the official setup-go library.
- Set up go modules cache.
- Replace staticcheck step with `make lint cover errcheck` in the test pipeline
- Removed all other settings from checkout code step except for fetch-depth.
- ~~Replaced Upload to codecov step with Upload to Github actions artifact.~~ Next to codecov step, github outputted artifacts are: Binary for the tested platform (linux-amd64) and the `coverage.out` file, which can be turned into HTML locally using `make cover-html`


### Release pipeline
- Run tests with `make pedantic`
- Use 1.18.x as golang version
- Use the official setup-go library.
- Set up go modules cache.

### Goreleaser

- Added before hook, runs `make tidy`
- Since this PR also adds the `cmd` directory, I changed the build config to work with multiple binaries instead of one.
- Added some recommended build options to enforce statically linked binaries.
- Previously goreleaser passed build-time args to `utils` module, this change passes them to the `main` instead.
- Added archive step, this creates a nice tar.gz file on the Releases page.
- Added milestones step, although not used currently, this is used for closing milestones if any based on the version number.
- ~~Added signs step for signing the output.~~ (Commented out until we add fingerprint and key in repo settings)

### Makefile

- Added new linting checks.
- Added build configuration for creating dev builds.
- Works with multiple binaries as well.

### File organization

- The contents of `cmd` are moved to `cmd/mach-composer`
- All internal libraries are moved to `internal/` folder.

### Linting fixes

- Some error handling.
- Use standard go naming.
- Formatting.
